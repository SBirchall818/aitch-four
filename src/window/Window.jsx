import React from 'react';
import { css } from 'aphrodite';
import styles from './window-styles';
import random from '../lib/random';
import Ball from '../ball/Ball';
import BallView from '../ball/BallView';
import iterateBallArray from '../lib/iterateBallArray';

export const RESIZE = 'resize';
export const TICK_MS = 50;

class Window extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      windowWidth: 0,
      windowHeight: 0,
      balls: [],
    };

    this.handleSizeChange = this.handleSizeChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.tick = this.tick.bind(this);
  }

  componentDidMount() {
    this.handleSizeChange();
    window.addEventListener(RESIZE, this.handleSizeChange);
    const intervalRef = setInterval(this.tick, TICK_MS);
    this.setState({ intervalRef });
  }

  componentWillUnmount() {
    window.removeEventListener(RESIZE, this.handleSizeChange);
    const { intervalRef } = this.state;
    clearInterval(intervalRef);
  }

  handleSizeChange() {
    this.setState({ windowWidth: window.innerWidth, windowHeight: window.innerHeight });
  }

  handleClick(e) {
    e.preventDefault();
    const { x, y } = e.nativeEvent;
    const { windowHeight, balls } = this.state;
    const h = windowHeight - y;

    const randomVelX = random(100);
    const randomVelH = random(100);
    this.setState({ balls: balls.concat(new Ball({ x, y, h }, { x: randomVelX, h: randomVelH })) });
  }

  tick() {
    const { balls, windowWidth, windowHeight } = this.state;
    const ballArrayNextState = iterateBallArray(balls, windowWidth, windowHeight);
    this.setState({ balls: ballArrayNextState });
  }

  render() {
    const { balls, windowWidth, windowHeight } = this.state;
    const ballList = balls.map((b) => <BallView key={b.key} x={b.pos.x} y={b.pos.y} />);
    return (
      <div
        onClick={this.handleClick}
        className={css(styles.root)}
        style={{ height: windowHeight, width: windowWidth }}
      >
        {ballList}
      </div>
    );
  }
}

export default Window;
