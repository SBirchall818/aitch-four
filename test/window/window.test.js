import React from 'react';
import { StyleSheetTestUtils } from 'aphrodite';
import { shallow } from 'enzyme';

import Window, { RESIZE } from '../../src/window/Window';

describe('Window :: ', () => {
  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();
  });

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  it('should exist', () => {
    expect(Window).toBeDefined();
  });

  it('should change state on window resize event', () => {
    const newHeight = 600;
    const newWidth = 500;

    const appComponent = shallow(<Window />);
    global.innerWidth = newWidth;
    global.innerHeight = newHeight;
    global.dispatchEvent(new Event(RESIZE));
    expect(appComponent.state().windowWidth).toEqual(newWidth);
    expect(appComponent.state().windowHeight).toEqual(newHeight);
  });

  it('adds a ball to App state when it is clicked', () => {
    const clickCoord = {x: 100, y: 120};
    const appComponent = shallow(<Window />);
    appComponent.find('div').simulate('click', { preventDefault() {}, nativeEvent: clickCoord });
    expect(appComponent.state().balls.length).toEqual(1);
    expect(appComponent.state().balls[0].pos.x).toEqual(clickCoord.x);
    expect(appComponent.state().balls[0].pos.y).toEqual(clickCoord.y);
  });
});