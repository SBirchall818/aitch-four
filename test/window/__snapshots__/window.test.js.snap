// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Window ::  should render Window defaults 1`] = `
<div
  className="root_k554e1"
  onClick={[Function]}
  style={
    Object {
      "height": 600,
      "width": 500,
    }
  }
/>
`;
